Let XML executables/nodes be "required" (like in ROS 1)

m-elwin opened the pull request:
Essentially on_exit="shutdown" is equivalent to ROS 1 required="true".

This feature is implemented using the python launchfile on_exit= mechanism.

Right now "shutdown" is the only action accepted by on_exit, but theoretically more "on_exit" actions could be added later.

Example:
<executable cmd="ls" on_exit="shutdown"/>

moriarty commented:
:+1: 
> This feature is implemented using the python launchfile on_exit= mechanism.

Linking related issue #174 and PR #179 added similar support for Python launch files via `on_exit=Shutdow()`

Py Example:

```py
    nodes_to_launch.append(
        Node(
            package="my_example_pkg",
            executable="example_node_main",
            name="example_node",
            ...
            on_exit=Shutdown(),
        )
    )
```

clalancette commented on launch/launch/actions/execute_process.py:
```diff
@@ -331,6 +331,16 @@ def parse(
+        if 'on_exit' not in ignore:
+            on_exit = entity.get_attr('on_exit', optional=True)
+            if on_exit is not None:
+                if on_exit == "shutdown":
```
Our style prefers single-quotes:

```suggestion
                if on_exit == 'shutdown':
```

clalancette reviewed (changes requested):
Thank you for the contribution.

Besides the fix listed inline, would you also please add this support to YAML?  That way we keep feature parity between XML and YAML.

m-elwin commented:
Thanks. Fixed the quotes and added a unit test for the yaml support. I think the code I added automatically covers both xml and yaml cases, but please let me know if I'm missing something.

clalancette reviewed (approved):
Thank you for iterating!  This looks great.  I'll run CI on it next.

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=20022)](http://ci.ros2.org/job/ci_linux/20022/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=14486)](http://ci.ros2.org/job/ci_linux-aarch64/14486/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20744)](http://ci.ros2.org/job/ci_windows/20744/)

23pointsNorth commented:
I can see this being backported to iron. Is there a plan to move this to humble as well?

emersonknapp commented:
@mergifyio backport humble
